src/transformers/models/patchtsmixer/modeling_patchtsmixer.py

```diff
@@ -0,0 +1,2361 @@
+
+    def _set_gradient_checkpointing(self, module, value=False):
+        if isinstance(module, (PatchTSMixerEncoder)):
+            module.gradient_checkpointing = value
+
+
+class PretrainHead(nn.Module):
+    """Pretrain head
+
+    Args:
+        num_patches (`int`): Number of patches to segment
+        patch_size (`int`, *optional*): Patch length. Defaults to 16.
+        input_size (`int`, *optional*): Number of input variables. Defaults to 1.
+        num_features (`int`, *optional*): Hidden feature size. Defaults to 16.
+        head_dropout (`float`, *optional*): Head Dropout rate. Defaults to 0.2.
+        mode (str, *optional*): Mixer Mode. Determines how to process the channels. Allowed values: flatten,
+            common_channel, mix_channel. In flatten, patch embedding encodes the patch information across all channels.
+            In common_channel mode, patch embedding is independent of channels (Channel Independece). In mix_channel,
+            we follow channel independence, but in addition to patch and feature mixing, we also do channel mixing.
+            Defaults to "common_channel".
+    """
+
+    def __init__(
+        self,
+        num_patches: int,
+        num_features: int = 16,
+        input_size: int = 1,
+        patch_size: int = 16,
+        head_dropout: float = 0,
+        mode: str = "common_channel",
+    ):
+        super().__init__()
+        self.mode = mode
+        self.patch_size = patch_size
+        self.input_size = input_size
+        self.num_patches = num_patches
+
+        if self.mode in ["common_channel", "mix_channel"]:
+            self.base_pt_block = nn.Sequential(
+                nn.Dropout(head_dropout),
+                nn.Linear(num_features, patch_size),
+            )
+        else:
+            self.base_pt_block = nn.Sequential(
+                nn.Dropout(head_dropout),
+                nn.Linear(num_features, patch_size * input_size),
+            )
+
+    def forward(self, x, y=None):
+        """
+        # flatten mode: [batch_size x num_patch x num_features] or
+            common_channel/mix_channel mode: [batch_size x n_vars x num_patch x num_features]
+
+        Output: z: [batch_size x n_vars x num_patch x patch_len]
+        """
+
+        if self.mode == "flatten":
+            x = self.base_pt_block(x)  # x: [batch_size x num_patch x n_vars*patch_size]
+            x = torch.reshape(
+                x, (x.shape[0], x.shape[1], self.patch_size, self.input_size)
+            )  # [batch_size x num_patch x patch_size x n_vars]
+            x = x.permute(0, 3, 1, 2)  # [batch_size x nvars x num_patch  x patch_len]
+            return x
+        elif self.mode in ["common_channel", "mix_channel"]:
+            forecast = self.base_pt_block(x)  # [batch_size x n_vars x num_patch x patch_size]
+            return forecast
+
+
+# TODO: add copied from after PatchTST master merge
+def positional_encoding(pe, learn_pe, q_len, d_model):
```

Comment:
same here, let's use copied from

Reply:
Yes this function looks to be exactly the same as here
```suggestion
# Copied from transformers.patchtst.modeling_patchtst.positional_encoding
def positional_encoding(pe, learn_pe, q_len, d_model):
```

Reply:
Ah ok there is already a comment!

Reply:
corrected